### ISSUE TITLE
Mismatch between cuda and cpu forward/backward projections 

### ISSUE THREAD

**Issue by josh-gree** [astra-toolbox/astra-toolbox#38]
This may not be the place to ask this and I am sorry if that is the case. I am trying to use astra to calculate forward and backward projections efficiently - however I seem to be seeing a large difference in the sinogram and backprojection operations for the same geometry when using the cuda rather than the cpu backend.

The following is the code I am running;

```
vol_geom = astra.create_vol_geom(256,256, -1,1,-1,1)
proj_geom = astra.create_proj_geom('parallel', 0.009, 256, np.linspace(0,np.pi,180,False))

#create cuda projector using vol and proj geom
proj_id = astra.create_projector('cuda',proj_geom,vol_geom)

cuda_sino_id,cuda_sino = astra.create_sino(P, proj_id)
cuda_backproject_id,cuda_backproject = astra.create_backprojection(cuda_sino,proj_id)

#create cpu projector using vol and proj geom
proj_id = astra.create_projector('line',proj_geom,vol_geom)

sino_id,sino = astra.create_sino(P, proj_id)
backproject_id,backproject = astra.create_backprojection(sino,proj_id)
```

if I look at plots of the sinogram and backprojections they differ by what appears to be a large scaling factor - my assumption is that the two should be approximately equal just that cuda should be more efficient?!

These are the plots I get;

![cudavscpu](https://cloud.githubusercontent.com/assets/8846667/14045314/047ed42e-f291-11e5-9eb0-5aee27e02b50.png)

The cuda sinogram seems to have values of around 100x smaller than the cpu version and the backprojection around 1000x larger. 

I am probably making a rookie mistake somewhere - any help would be appreciated!!

**Comment by dmpelt** [comment]
This is a good place to ask this kind of question, please don't hesitate to create an issue or send us an email if you have other issues in the future.

I don't know the exact details of the specific implementations of both backends, but could you try to change the first two lines to:

```
vol_geom = astra.create_vol_geom(256,256)
proj_geom = astra.create_proj_geom('parallel', 1.0, 256, np.linspace(0,np.pi,180,False))
```

In this way, the pixels of the reconstruction image are defined as pixels with a width and height of 1, which is the usual way to use the ASTRA toolbox. It could be that different backends support non `1x1` pixels differently (some backends might not support it at all), which may cause the problems you are seeing. If the detector pixels have a different size than the image pixels, you can change the second argument of `astra.create_proj_geom` accordingly (in the above example, you could use `0.009*128=1.152` to get results similar to your example).

In general, I would recommend to always use `1x1` pixels in the volume geometry, and scale the detector pixel size in the projection geometry accordingly.

**Comment by wjp** [comment]
Due to oversights in the past, there are indeed some differences in output scaling between different CPU projectors, and between CPU and GPU code. The `strip` projector should at least for the FP have the same scaling behaviour as the GPU code, but the `line` projector scales differently as you have noticed. The backprojection is a different matter again, sadly.

We've been holding off on fixing these inconsistencies as it would unexpectedly change the behaviour of existing code, but we plan to fix this when we release ASTRA 2.0. (Which is tentatively planned for sometime in 2016.)

**Comment by adler-j** [comment]
I'll do a short plug for [odl](https://github.com/odlgroup/odl/) here. In ODL we have tried to make projectors and the likes match the mathematical operators that they represent, in this case the ray transform. 

Trying your example ([source code here](https://gist.github.com/adler-j/e9eddcf4ef919042a72178e523429142), modified from [this example](https://github.com/odlgroup/odl/blob/master/examples/tomo/ray_trafo_parallel_2d.py)) we get the expected result:

![results](http://i.imgur.com/OaLcVBT.png)

All of this is (of course) done using the ASTRA projectors!

**Comment by jakobsj** [comment]
I maybe observing just this issue when trying to compare CGLS and FBP reconstructions using GPU code and the fanflat geometry. I thought I had figured out the scaling factor, but it's not exactly right. But it might also be a constant offset, as the FBP reconstruction has negative values of almost constant magnitude, where it should be zero. Matlab's ifanbeam does not produce the same offset. I'll post my code below. Do you know what the correct scaling factor is -- or if it is not a scaling factor issue but something else?

**Comment by jakobsj** [comment]
Here's the resulting figure, with true phantom in black, CGLS in blue, ASTRA FBP in red and Matlab's ifanbeam in magenta.
![fbp_scaling](https://cloud.githubusercontent.com/assets/9976072/16286140/675b4a32-38d2-11e6-88d8-12f26498a62c.png)

I'm using ASTRA 1.7.1 on linux. Here's the code:

N = 512;
sc = 10000;
cd = 0; 
na = 2000;
mag = (sc+cd)/sc;
vol_geom = astra_create_vol_geom(N,N);
proj_geom = astra_create_proj_geom('fanflat', mag, N, linspace2(0,2*pi,na), sc,cd);

% Create shepp logan sinogram data
P = phantom(N);
[sinogram_id, sinogram] = astra_create_sino_gpu(P, proj_geom, vol_geom);
figure(1); imshow(P, []);
colorbar
drawnow

figure(2); imshow(sinogram, []);
colorbar
drawnow

rec_id = astra_mex_data2d('create', '-vol', vol_geom);

% create configuration 
cfg = astra_struct('FBP_CUDA');
cfg.ReconstructionDataId = rec_id;
cfg.ProjectionDataId = sinogram_id;
cfg.FilterType = 'Ram-Lak';

% Create and run the algorithm object from the configuration structure
alg_id = astra_mex_algorithm('create', cfg);
astra_mex_algorithm('run', alg_id);

% Get the FBP result
cor_fac = proj_geom.DistanceOriginSource^2;
rec_fbp = cor_fac*astra_mex_data2d('get', rec_id);
figure(3); imshow(rec_fbp, []);
colorbar
drawnow

% Now do CGLS recon
rec_id2 = astra_mex_data2d('create','-vol',vol_geom);

cfg2 = astra_struct('CGLS_CUDA');
cfg2.ReconstructionDataId = rec_id2;
cfg2.ProjectionDataId = sinogram_id;

% Create and run the algorithm object from the configuration structure
alg_id2 = astra_mex_algorithm('create', cfg2);
astra_mex_algorithm('run', alg_id2, 100);

% Get the CGLS result
rec_cgls = astra_mex_data2d('get', rec_id2);
figure(4); imshow(rec_cgls, []);
colorbar
drawnow

% Clean up. Note that GPU memory is tied up in the algorithm object,
% and main RAM in the data objects.
astra_mex_algorithm('delete', alg_id);
astra_mex_data2d('delete', rec_id);
astra_mex_data2d('delete', sinogram_id);
astra_mex_data2d('delete', rec_id2);

%% Try ifanbeam
rec_ifanbeam = ifanbeam(sinogram',sc, 'FanSensorGeometry', 'line','outputsize',N);

%% Plot profiles
figure
plot(rec_cgls(:,end/2))
hold on
plot(rec_fbp(:,end/2),'--r')
plot(P(:,end/2),'-k')
plot(rec_ifanbeam(:,end/2),'-m')
hold off

**Comment by wjp** [comment]
The factor `proj_geom.DistanceOriginSource^2` is indeed the right one here. I'll take a closer look att the offset to see where that's coming from.

**Comment by wjp** [comment]
The offset seems to be caused by Matlab computing the ramp filter in the spatial domain, while we build it directly in the frequency domain. We should add this option too (or maybe switch to it entirely).

**Comment by jakobsj** [comment]
Thanks very much for looking into it and for confirming that the factor is correct. Regarding offset, are you sure Matlab computes the filter in the spatial domain? The notes under help for both ifanbeam and iradon, which ifanbeam calls, state that "The filter is designed directly in the frequency domain and then multiplied by the FFT of the projections", and the code does seem to agree with that.

**Comment by wjp** [comment]
Yes, see the `designFilter` function in `iradon.m`. It creates `filtImpResp` in the spatial domain, and then takes its FFT to get the frequency filter. (Matlab 2015a)

Until a couple of years ago (2011? 2012? I can't easily check.), Matlab did construct its ramp filter directly in the frequency domain, though.

**Comment by wjp** [comment]
Small addition to the scaling factor: in general it is `proj_geom.DistanceOriginSource * (proj_geom.DistanceOriginSource + proj_geom.DistanceOriginDetector)`.

**Comment by jakobsj** [comment]
Thanks for the more general scaling factor. I checked in Matlab R2014b and you are right, it is being created in the spatial domain, then FFT'ed - I suppose they didn't update the documentation then. 

The offset does appear constant, so the difference may just be due to different DC components used? So as of yet, there's no built-in option in ASTRA to do this? I think I saw somewhere in the source, that there's the option to pass in your own filter, maybe using the 'rprojection' or 'projection' filtertype or something like that (can't find it again now), so perhaps could manually feed in the same one MATLAB constructs to get rid of the offset?

**Comment by wjp** [comment]
Yes, you can pass your own filter using the `projection` (frequency domain) or `rprojection` (spatial domain) types.

For `rprojection` the size of the spatial filter must be the smallest power of two at least twice the width of the sinogram. (So 1024 in your example.)

For `projection` the size of the frequency filter must be half of that, plus 1, and the values represent the real part of the non-negative half of the ramp. (So 513 in your example.)

The actual values go into a data2d object of sinogram-type, and you pass its ID as FilterSinogramId to the FBP_CUDA algorithm.

For example, if you take `filtImpResp` from matlab's `designFilter`:

    filter_geom = astra_create_proj_geom('fanflat', 1.0, 1024, [0], 1, 0); % dummy values; only size matters
    filter_id = astra_mex_data2d('create', '-sino', filter_geom, 2*ifftshift(filtImpResp));

    % create configuration 
    cfg = astra_struct('FBP_CUDA');
    cfg.ReconstructionDataId = rec_id;
    cfg.ProjectionDataId = sinogram_id;
    cfg.FilterSinogramId = filter_id;
    cfg.FilterType = 'rprojection';

**Comment by wjp** [comment]
If you're interested, the code to have ASTRA directly generate a filter in the spatial domain is at #59 .

**Comment by jakobsj** [comment]
Thanks very much for putting this together. I've tried out #59 but I get an error. I cloned the whole thing, installed as per usual instructions with no issues, but when I try to run the example code above (or the supplied samples s001...s004, but not s005) I get the error pasted below (for s001). Does it work for you/any idea what's wrong?

>> s001_sinogram_par2d
Unexpected Standard exception from MEX file.
What() is:std::bad_alloc
..

Error in astra_mex_data2d (line 23)
    [varargout{:}] = astra_mex_data2d_c(varargin{:});

Error in astra_create_sino_gpu (line 36)
sino_id = astra_mex_data2d('create','-sino', proj_geom, 0);

Error in s001_sinogram_par2d (line 26)
[sinogram_id, sinogram] = astra_create_sino_gpu(P, proj_geom, vol_geom);

**Comment by jakobsj** [comment]
Sorry, here it is:
```Matlab
>> s001_sinogram_par2d
Unexpected Standard exception from MEX file.
What() is:std::bad_alloc
..

Error in astra_mex_data2d (line 23)
    [varargout{:}] = astra_mex_data2d_c(varargin{:});

Error in astra_create_sino_gpu (line 36)
sino_id = astra_mex_data2d('create','-sino', proj_geom, 0);

Error in s001_sinogram_par2d (line 26)
[sinogram_id, sinogram] = astra_create_sino_gpu(P, proj_geom, vol_geom);
 
```

**Comment by AceCoooool** [comment]
> Due to oversights in the past, there are indeed some differences in output scaling between different CPU projectors, and between CPU and GPU code. The `strip` projector should at least for the FP have the same scaling behaviour as the GPU code, but the `line` projector scales differently as you have noticed. The backprojection is a different matter again, sadly.
> 
> We've been holding off on fixing these inconsistencies as it would unexpectedly change the behaviour of existing code, but we plan to fix this when we release ASTRA 2.0. (Which is tentatively planned for sometime in 2016.)

@wjp Is there any progress in this mismatch?
